biogrid/libs/biogrid-simulator/src/lib/biogrid/biogrid-class.ts

```diff
@@ -5,62 +5,176 @@ import {
+    // TODO constants
+    this.smallBatteries = this.createBatteries(smallBatteryPositions, SMALL_BATTERY.DEFAULT_START_ENERGY, SMALL_BATTERY.MAX_CAPACITY, GRID_ITEM_NAMES.SMALL_BATTERY);
+    this.largeBatteries = this.createBatteries(largeBatteryPositions, LARGE_BATTERY.DEFAULT_START_ENERGY, LARGE_BATTERY.MAX_CAPACITY, GRID_ITEM_NAMES.LARGE_BATTERY);
+
+    // Towns
+    this.town = town;
+
+    // Enery Source
+    // TODO implement the solar panels
+    const solarPanelPositions = this.createGridItemPositions(town.getTownSize(), opts.numberOfSolarPanels);
+    this.solarPanels = this.createSolarPanels(solarPanelPositions);
+
+    this.state = new BiogridState(this.createGridItems());
+
   }
 
-  getSystemState(): BiogridState {
+  private createGridItems(): GridItem[] {
+    return [
+      ...this.smallBatteries,
+      ...this.largeBatteries,
+      ...this.town.getEnergyUsers(),
+      ...this.solarPanels,
+    ]
+  }
+
+  getSystemState() {
     return this.state;
   }
 
+  getJsonGraphDetails() {
+    this.state.getJsonGraph();
+  }
+
+  private createBatteries(positions: ItemPosition[], initEnergy: Energy, maxCapacity: Energy, name: string): Battery[] {
+    return positions.map((position, index) => new BioBattery(position.x, position.y, `${name}-${index}`, initEnergy, maxCapacity));
+  }
+
+  private createSolarPanels(positions: ItemPosition[]): EnergySource[] {
+    return positions.map((position, index) => new BioEnergySource(position.x, position.y, `${GRID_ITEM_NAMES.SOLAR_PANEL}-${index}`, SOLAR_PANEL.DEFAULT_INITIAL_ENERGY, SOLAR_PANEL.MIN_CAPACITY))
+  }
+
   takeAction(action: GridAction) {
-    // TODO implement actions taken later
-    return;
+    // RETURN a new BiogridState
+    const allSupplyingPaths = action.getSupplyingPaths()
+
+    const clonedGraph = this.state.cloneStateGraph();
+
+    for (const supplyPath in allSupplyingPaths) {
+      const oldGridItem = this.state.getGridItem(supplyPath);
+      const supplyingGridItem = this.state.getGridItem(allSupplyingPaths[supplyPath]);
+      const typeOldGridItem = this.getGridItemType(oldGridItem);
+      if (typeOldGridItem === GRID_ITEM_NAMES.ENERGY_USER) {
+        const energyUser = oldGridItem as Building;
+        const energyUserReq = energyUser.MaxCapacity - energyUser.getEnergyInJoules();
+        const typeSupplyingGridItem = this.getGridItemType(supplyingGridItem);
+        if (typeSupplyingGridItem === GRID_ITEM_NAMES.LARGE_BATTERY || typeSupplyingGridItem === GRID_ITEM_NAMES.SMALL_BATTERY) {
+          const battery = supplyingGridItem as BioBattery;
+          battery.supplyPower(energyUserReq);
+          clonedGraph.setNode(battery.name, battery);
+        } else if (typeSupplyingGridItem === GRID_ITEM_NAMES.SOLAR_PANEL) {
+          const solarpanel = supplyingGridItem as BioEnergySource;
+          solarpanel.supplyPower(energyUserReq);
+          clonedGraph.setNode(solarpanel.name, solarpanel);
+        } else {
```

Comment:
nit: remove else and continue

Reply:
i cannot remove it because a node which has not been changed, will cause an error when I request to start charging it or change it
check lines after the `if..else`